aten/src/ATen/ParallelThreadPoolNative.cpp

```diff
@@ -0,0 +1,77 @@
+#if AT_PARALLEL_OPENMP
+#include <ATen/Parallel.h>
+#include <ATen/PTThreadPool.h>
+
+#include <atomic>
+#include <thread>
+
+namespace at {
+
+namespace {
+const int NOT_SET = -1;
+const int CONSUMED = -2;
+
+// Number of inter-op threads set by the user;
+// Atomic transitions:
+// NOT_SET -> (atomic) -> positive value -> (atomic) -> CONSUMED
+// (CONSUMED - thread pool is initialized)
+// or
+// NOT_SET -> (atomic) -> CONSUMED
+std::atomic<int> num_interop_threads{NOT_SET};
+
+// thread pool global instance is hidden,
+// users should use at::launch ang get/set_num_interop_threads interface
+TaskThreadPoolBase& get_pool() {
+  static std::shared_ptr<TaskThreadPoolBase> pool =
+      ThreadPoolRegistry()->Create(
+          "C10",
+          /* device_id */ 0,
+          /* pool_size */ num_interop_threads.exchange(CONSUMED),
+          /* create_new */ true);
+  return *pool;
+}
+
+// Factory function for ThreadPoolRegistry
+std::shared_ptr<TaskThreadPoolBase> create_c10_threadpool(
+    int device_id,
+    int pool_size,
+    bool create_new) {
+  // For now, the only accepted device id is 0
+  // for the JIT inter-op pool (CPU),
+  AT_CHECK(device_id == 0);
+  // Create new thread pool
+  AT_CHECK(create_new);
+  return std::make_shared<PTThreadPool>(pool_size);
+}
+
+} // namespace
+
+C10_REGISTER_CREATOR(ThreadPoolRegistry, C10, create_c10_threadpool);
+
+void set_num_interop_threads(int nthreads) {
+  AT_CHECK(nthreads > 0, "Expected positive number of threads");
+
+  int no_value = NOT_SET;
+  AT_CHECK(num_interop_threads.compare_exchange_strong(no_value, nthreads),
+      "Error: cannot set number of interop threads after parallel work "
+      "has started or set_num_interop_threads called");
+}
+
+int get_num_interop_threads() {
```

Comment:
Why is this moving from the more general spot https://github.com/pytorch/pytorch/pull/20050/files#diff-c2f99b959224fab235eecd2cc4e696b8R160 to here?

Reply:
because the implementation of `get_num_interop_threads` depends on how we implement inter-op parallelism - we have two choices - native thread pool (here) and (potentially) TBB

Reply:
so ParallelThreadPoolNative.cpp specifically implements inter-op parallelism using a native thread pool